File: Source/Init/Difficulty/DifficultyOption.cs
using System.Collections.Generic;

/// <summary>
/// A single difficulty tier that can be voted on. Pure data - the dialog and
/// button wiring live in <see cref="Difficulty"/>, which owns a
/// SelectionDialog and maps each of its buttons back to one of these.
/// </summary>
public class DifficultyOption
{
    public static List<DifficultyOption> Options { get; } = new List<DifficultyOption>();

    public string Name { get; }
    public int Value { get; }
    public string Color { get; }
    public int TallyCount { get; set; }

    public DifficultyOption(string name, int value, string color)
    {
        Name = name;
        Value = value;
        Color = color;
        Options.Add(this);
    }

    public override string ToString()
    {
        return $"{Color}{Name}|r";
    }

    public static void Initialize()
    {
        Options.Clear();
        new DifficultyOption("Progressive (3 Rounds)", (int)DifficultyLevel.Progressive, Colors.COLOR_GREEN);
        new DifficultyOption("Normal", (int)DifficultyLevel.Normal, Colors.COLOR_YELLOW);
        new DifficultyOption("Hard", (int)DifficultyLevel.Hard, Colors.COLOR_RED);
        new DifficultyOption("Impossible", (int)DifficultyLevel.Impossible, Colors.COLOR_DARK_RED);
        new DifficultyOption("Nightmare", (int)DifficultyLevel.Nightmare, Colors.COLOR_PURPLE);
    }
}

public enum DifficultyLevel
{
    Normal = 4,
    Hard = 6,
    Impossible = 9,
    Nightmare = 11,
    Progressive = 99
}

File: src/Init/Difficulty/DifficultyOption.ts
import { Colors } from 'src/Utility/Colors/Colors'
import { Dialog, DialogButton } from 'w3ts'

export class DifficultyOption {
    public static Options: DifficultyOption[] = []
    public static DifficultyChoosing: Dialog = Dialog.create()!
    private static _OptionCount: number = -1
    public name: string
    public Value = 0
    public Color: string
    public Button: DialogButton
    public TallyCount = 0

    constructor(name: string, value: number, color: string) {
        this.name = name
        this.Value = value
        this.Color = color
        this.Button = this.AddButton()
        DifficultyOption.Options.push(this)
    }

    public AddButton(): DialogButton {
        return DifficultyOption.DifficultyChoosing.addButton(
            `${this.Color}${this.name}|r`,
            ++DifficultyOption._OptionCount
        )!
    }

    public toString(): string {
        return `${this.Color}${this.name}|r`
    }

    public static Initialize = () => {
        DifficultyOption.Options = []
        new DifficultyOption('Normal', DifficultyLevel.Normal, Colors.COLOR_YELLOW)
        new DifficultyOption('Hard', DifficultyLevel.Hard, Colors.COLOR_RED)
        new DifficultyOption('Impossible', DifficultyLevel.Impossible, Colors.COLOR_DARK_RED)
        new DifficultyOption('Nightmare', DifficultyLevel.Nightmare, Colors.COLOR_PURPLE)
    }
}

export enum DifficultyLevel {
    Normal = 4,
    Hard = 6,
    Impossible = 9,
    Nightmare = 11,
}
